### PR TITLE
cups-brother-mfcl2800dw: init at 4.1.0-1

### LINKED ISSUE #311925: Package request: cups-brother-mfcl2800dw
**Project description**

CUPS driver for the Brother MFC-L2800DW

**Metadata**

* homepage URL: https://support.brother.com/g/b/producttop.aspx?c=gb&lang=en&prod=mfcl2800dw_eu_as

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by luftmensch-luftmensch**
## Description of changes
Closes #311925

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by luftmensch-luftmensch** [comment]
```
> nix build .#cups-brother-mfcl2800dw
> tree result/

 result
├──  etc
│  └──  opt
│     └──  brother
│        └──  Printers
│           └──  MFCL2800DW
│              └──  inf
├──  lib
│  └──  cups
│     └──  filter
│        └──  brother_lpdwrapper_MFCL2800DW -> ../../../opt/brother/Printers/MFCL2800DW/lpd/lpdfilter
├──  opt
│  └──  brother
│     └──  Printers
│        └──  MFCL2800DW
│           ├──  cupswrapper
│           │  ├──  .lpdwrapper-wrapped
│           │  ├──  .paperconfigml2-wrapped
│           │  ├──  brother-MFCL2800DW-cups-en.ppd
│           │  ├──  Copying
│           │  ├──  lpdwrapper
│           │  └──  paperconfigml2
│           ├──  inf
│           │  ├──  .setupPrintcap-wrapped
│           │  ├──  brMFCL2800DWfunc
│           │  ├──  brMFCL2800DWrc
│           │  └──  setupPrintcap
│           ├──  lpd
│           │  ├──  inf
│           │  ├──  x86_64
│           │  │  ├──  .brprintconflsr3-wrapped
│           │  │  ├──  .rawtobr3-wrapped
│           │  │  ├──  brprintconflsr3
│           │  │  └──  rawtobr3
│           │  ├──  .lpdfilter-wrapped
│           │  ├──  brprintconflsr3 -> x86_64/brprintconflsr3
│           │  ├──  lpdfilter
│           │  └──  rawtobr3 -> x86_64/rawtobr3
│           ├──  LICENSE_ENG.txt
│           └──  LICENSE_JPN.txt
├──  share
│  └──  cups
│     └──  model
│        └──  brother-MFCL2800DW-cups-en.ppd -> ../../../opt/brother/Printers/MFCL2800DW/cupswrapper/brother-MFCL2800DW-cups-en.ppd
├──  usr
│  └──  share
│     └──  doc
└──  var
   └──  spool
      └──  lpd
         └──  MFCL2800DW
```

**Comment by Pandapip1** [comment]
Result of `nixpkgs-review pr 311949` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cups-brother-mfcl2800dw</li>
  </ul>
</details>

**Review by Pandapip1 — APPROVED** [review]
LGTM.

File: pkgs/by-name/cu/cups-brother-mfcl2800dw/package.nix [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        # bundled scripts don't understand the arch subdirectories for some reason
+        ln -s \
+          "$out/opt/brother/Printers/MFCL2800DW/lpd/${stdenv.hostPlatform.linuxArch}/"* \
+          "$out/opt/brother/Printers/MFCL2800DW/lpd/"
+
+        # Fix global references and replace auto discovery mechanism with hardcoded values
+        substituteInPlace $out/opt/brother/Printers/MFCL2800DW/lpd/lpdfilter \
+          --replace /opt "$out/opt" \
+          --replace "my \$BR_PRT_PATH =" "my \$BR_PRT_PATH = \"$out/opt/brother/Printers/MFCL2800DW\"; #" \
+          --replace "PRINTER =~" "PRINTER = \"MFCL2800DW\"; #"
```
Comment (SuperSandro2000) [inline pkgs/by-name/cu/cups-brother-mfcl2800dw/package.nix:73]:
```suggestion
          --replace-fail /opt "$out/opt" \
          --replace-fail "my \$BR_PRT_PATH =" "my \$BR_PRT_PATH = \"$out/opt/brother/Printers/MFCL2800DW\"; #" \
          --replace-fail "PRINTER =~" "PRINTER = \"MFCL2800DW\"; #"
```

File: pkgs/by-name/cu/cups-brother-mfcl2800dw/package.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+        # Make sure all executables have the necessary runtime dependencies available
+        find "$out" -executable -and -type f | while read file; do
+          wrapProgram "$file" --prefix PATH : "${lib.makeBinPath runtimeDeps}"
+        done
+
+        # Symlink filter and ppd into a location where CUPS will discover it
+        mkdir -p $out/lib/cups/filter
+        mkdir -p $out/share/cups/model
```
Comment (SuperSandro2000) [inline pkgs/by-name/cu/cups-brother-mfcl2800dw/package.nix:82]:
```suggestion
        mkdir -p $out/lib/cups/filter $out/share/cups/model
```

File: pkgs/by-name/cu/cups-brother-mfcl2800dw/package.nix (hunk 1 above)
Comment (luftmensch-luftmensch) [reply pkgs/by-name/cu/cups-brother-mfcl2800dw/package.nix]: Applied suggestion!

**Review by Pandapip1 — APPROVED** [review]
Still LGTM

**Pull request merged**